### PR TITLE
csi-wrapper: add example for AWS EBS CSI driver

### PR DESCRIPTION
**Pull request by snir911**
and how to add the peer-pods CSI wrapper to it

Fixes: #1363

### CODE REVIEW

File: volumes/csi-wrapper/examples/aws/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+kubectl apply -f rbac-ebs-csi-wrapper-runner.yaml
+kubectl apply -f rbac-ebs-csi-wrapper-podvm.yaml
+```
+3. Patch the EBS CSI Driver:
+```
+kubectl patch deploy ebs-csi-controller -n kube-system --patch-file patch-controller.yaml
+kubectl -n kube-system delete replicaset -l app=ebs-csi-controller
+kubectl patch ds ebs-csi-node -n kube-system --patch-file patch-node.yaml
+```
+4. Verify the pods are running (each pod should contain one more container than before the patching):
```
Comment (bpradipt) [inline volumes/csi-wrapper/examples/aws/README.md:52]: nit
```suggestion
4. Verify the pods are running (each pod should contain an additional container):
```

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @snir911

File: volumes/csi-wrapper/examples/aws/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## Prerequisites
+
+* Running Kubernetes cluster running on AWS Version >= 1.20
+
+* Peer-Pods is [deployed](../../../../aws/README.md)
+
+## AWS EBS CSI Driver Installtion
+
+**NOTE:** the following is just a basic example, follow official [installtion instructions](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/install.md) for advanced configuration.
```
Comment (wainersm) [inline volumes/csi-wrapper/examples/aws/README.md:11]: s/installtion/installation/

File: volumes/csi-wrapper/examples/aws/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# AWS EBS CSI Wrapper for Peer Pod Storage
+
+## Prerequisites
+
+* Running Kubernetes cluster running on AWS Version >= 1.20
```
Comment (wainersm) [inline volumes/csi-wrapper/examples/aws/README.md:5]: Maybe "Running Kubernetes cluster (version >= 1.20) on AWS" ?

**Comment by wainersm** [comment]
hi @snir911 ! Mind if I hold this PR until I give it a try? Never used CSI with peer pods...so I'd like to use that opportunity to learn :D

**Comment by wainersm** [comment]
hi @snir911 ,

I finally got some free time to try it out, but the pod app is in crashloopback status:

```
[wmoschet@wainer-laptop aws]$ kubectl describe pod/app
Name:         app
Namespace:    default
Priority:     0
Node:         ip-10-0-0-91.ec2.internal/10.0.0.91
Start Time:   Tue, 05 Sep 2023 15:19:53 -0300
Labels:       <none>
Annotations:  <none>
Status:       Running
IP:           10.0.0.104
IPs:
  IP:  10.0.0.104
Containers:
  app:
    Container ID:  containerd://5cedbe174c8a217b5d11a7d9d2becd420164359a79f5737d8121276442d97f4b
    Image:         centos
    Image ID:      [Encrypted]
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
    Args:
      -c
      while true; do echo $(date -u) >> /data/out.txt; sleep 5; done
    State:          Running
      Started:      Tue, 05 Sep 2023 15:21:14 -0300
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /data from persistent-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-p544x (ro)
  csi-podvm-node-driver:
    Container ID:   containerd://0a2b3f0059772280a4b126669b0e781077e2b62d9bd7181b153f43734305f112
    Image:          public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.21.0
    Image ID:       [Encrypted]
    Port:           9808/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 05 Sep 2023 15:22:02 -0300
      Finished:     Tue, 05 Sep 2023 15:22:18 -0300
    Ready:          False
    Restart Count:  2
    Environment:
      KUBE_NODE_NAME:   (v1:spec.nodeName)
      CSI_NODE_NAME:    (v1:spec.nodeName)
    Mounts:
      /tmp from plugin-dir (rw)
      /var/lib/kubelet from kubelet-data-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-p544x (ro)
  csi-podvm-wrapper:
    Container ID:  containerd://3846b13d62ce119d18e5c71d45fba5e991e39f3c4e6151a1374173f7a40d57ea
    Image:         quay.io/confidential-containers/csi-podvm-wrapper:latest
    Image ID:      [Encrypted]
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/bin/csi-podvm-wrapper
    Args:
      --v=2
      --endpoint=/tmp/csi-podvm-wrapper.sock
      --target-endpoint=/tmp/csi.sock
      --namespace=kube-system
    State:          Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Tue, 05 Sep 2023 15:22:19 -0300
      Finished:     Tue, 05 Sep 2023 15:22:21 -0300
    Last State:     Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Tue, 05 Sep 2023 15:21:47 -0300
      Finished:     Tue, 05 Sep 2023 15:21:48 -0300
    Ready:          False
    Restart Count:  3
    Environment:
      BINARY:          csi-podvm-wrapper
      POD_NAME:        app (v1:metadata.name)
      POD_NAME_SPACE:  default (v1:metadata.namespace)
      POD_UID:          (v1:metadata.uid)
      POD_NODE_NAME:    (v1:spec.nodeName)
    Mounts:
      /tmp from plugin-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-p544x (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  persistent-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  ebs-claim
    ReadOnly:   false
  kubelet-data-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet
    HostPathType:  Directory
  plugin-dir:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  kube-api-access-p544x:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              katacontainers.io/kata-runtime=true
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Warning  FailedScheduling        2m40s              default-scheduler        0/1 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod..
  Normal   Scheduled               2m35s              default-scheduler        Successfully assigned default/app to ip-10-0-0-91.ec2.internal
  Normal   SuccessfulAttachVolume  2m33s              attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-8ad28226-bdf0-4933-8401-bbddbc488cb1"
  Normal   Pulling                 79s                kubelet                  Pulling image "centos"
  Normal   Pulled                  75s                kubelet                  Successfully pulled image "centos" in 4.645341073s (4.645349227s including waiting)
  Normal   Created                 75s                kubelet                  Created container app
  Normal   Started                 74s                kubelet                  Started container app
  Normal   Pulled                  72s                kubelet                  Successfully pulled image "public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.21.0" in 1.825911694s (1.825924865s including waiting)
  Normal   Pulled                  68s                kubelet                  Successfully pulled image "quay.io/confidential-containers/csi-podvm-wrapper:latest" in 2.987932919s (2.987941026s including waiting)
  Normal   Pulled                  59s                kubelet                  Successfully pulled image "quay.io/confidential-containers/csi-podvm-wrapper:latest" in 231.71683ms (231.740711ms including waiting)
  Normal   Started                 59s (x2 over 67s)  kubelet                  Started container csi-podvm-wrapper
  Normal   Created                 59s (x2 over 68s)  kubelet                  Created container csi-podvm-wrapper
  Normal   Created                 55s (x2 over 72s)  kubelet                  Created container csi-podvm-node-driver
  Normal   Pulled                  55s                kubelet                  Successfully pulled image "public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.21.0" in 164.242194ms (164.251044ms including waiting)
  Normal   Started                 55s (x2 over 71s)  kubelet                  Started container csi-podvm-node-driver
  Normal   Pulling                 55s (x2 over 74s)  kubelet                  Pulling image "public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.21.0"
  Warning  BackOff                 53s (x3 over 55s)  kubelet                  Back-off restarting failed container csi-podvm-wrapper in pod app_default(1cbf1bfb-036d-4ded-a600-e4d15ba8fded)
  Normal   Pulling                 42s (x3 over 71s)  kubelet                  Pulling image "quay.io/confidential-containers/csi-podvm-wrapper:latest"
  Normal   Pulled                  42s                kubelet                  Successfully pulled image "quay.io/confidential-containers/csi-podvm-wrapper:latest" in 432.751745ms (432.811723ms including waiting)
[wmoschet@wainer-laptop aws]$ kubectl get pods
NAME   READY   STATUS             RESTARTS      AGE
app    1/3     CrashLoopBackOff   5 (13s ago)   2m43s
```

The PVC claim is `Bound`:
```
[wmoschet@wainer-laptop aws]$ kubectl get pvc ebs-claim
NAME        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
ebs-claim   Bound    pvc-8ad28226-bdf0-4933-8401-bbddbc488cb1   4Gi        RWO            ebs-sc         4m32s
```

The EBS pods running:
```
[wmoschet@wainer-laptop aws]$ kubectl get pods -n kube-system -l app.kubernetes.io/name=aws-ebs-csi-driver
NAME                                  READY   STATUS    RESTARTS   AGE
ebs-csi-controller-6cdc9cd849-qprbt   7/7     Running   0          10m
ebs-csi-node-qlzd4                    4/4     Running   0          10m
```

I'm running in a VPC with inbound/outbound ports restricted, basically I'm just opening https, ssh, icmp and caa proxy ports. Might this be the problem?

File: volumes/csi-wrapper/examples/aws/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+kubectl apply -f rbac-ebs-csi-wrapper-runner.yaml
+kubectl apply -f rbac-ebs-csi-wrapper-podvm.yaml
+```
+3. Patch the EBS CSI Driver:
+```
+kubectl patch deploy ebs-csi-controller -n kube-system --patch-file patch-controller.yaml
+kubectl -n kube-system delete replicaset -l app=ebs-csi-controller
+kubectl patch ds ebs-csi-node -n kube-system --patch-file patch-node.yaml
+```
+4. Verify the pods are running (each pod should contain an additional container):
```
Comment (wainersm) [inline volumes/csi-wrapper/examples/aws/README.md:52]: I didn't understand `each pod should contain an additional container`. I see the pods running and I assumed it is fine.

File: volumes/csi-wrapper/examples/aws/dynamic-provisioning/pod.yaml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    env:
+      - name: KUBE_NODE_NAME
+        valueFrom:
+          fieldRef:
+            fieldPath: spec.nodeName
+      - name: CSI_NODE_NAME
+        valueFrom:
+          fieldRef:
+            fieldPath: spec.nodeName
+     # set when IMDS isn't accessible from the podvm
```
Comment (wainersm) [inline volumes/csi-wrapper/examples/aws/dynamic-provisioning/pod.yaml:38]: remove this block?

**Comment by wainersm** [comment]
One last comment: it would be nice to have instructions to uninstall the csi-wrapper.

File: volumes/csi-wrapper/examples/aws/README.md (hunk 4 above)
Comment (snir911) [reply volumes/csi-wrapper/examples/aws/README.md:52]: hmm,  for example if previously pod showed 3/3 containers, after patching it should show 4/4, suggestions for better wording are welcomed

File: volumes/csi-wrapper/examples/aws/dynamic-provisioning/pod.yaml (hunk 5 above)
Comment (snir911) [reply volumes/csi-wrapper/examples/aws/dynamic-provisioning/pod.yaml:38]: I'd leave it as IIUC in some cases IMDS might be disabled

**Comment by bpradipt** [comment]
@snir911 @wainersm what is pending on this PR ?

**Comment by snir911** [comment]
@wainersm ok to merge on your side? I'll add instructions to delete once I'll have this setup again and I'll be able to verify, for now, AFAIR it should be specified in the official aws driver instructions and nothing is special for deleting the wrapper

**Review by wainersm — APPROVED** [review]

**Pull request merged**